TACC/Core-Portal
# task/FP-1760 - apcd project membership

## Overview

Adds an a 'userlink' feature to the Project Membership onboarding step that forwards a user to a request page upon clicking "Request Access" (instead of sending an RT ticket) and will retry checking membership when viewing the onboarding splash page

## Related

* [FP-1760](https://jira.tacc.utexas.edu/browse/FP-1760)

## Changes

Back End:

- project_membership_step - checks for membership and sends a `userlink` data structure to front end for rendering a request access link
- onboarding/api/views:get_user_onboarding - if a step is encountered that has the 'retry' flag set in its settings and it has NOT completed and is NOT pending (in other words, has failed or is in some other type of wait state) then it will attempt to process the step and trigger the rest of onboarding if the step is successful

Front End:

- When rendering a step in the "userwait" state, if the step data contains a `userlink` data structure it will render the Request Access button as a link

## Testing

0. Configure ngrok!
1. Delete your local user
2. Set this as your onboarding configuration's first step in `settings_custom.py`

```
_PORTAL_USER_ACCOUNT_SETUP_STEPS = [
            {
        'step': 'portal.apps.onboarding.steps.project_membership.ProjectMembershipStep',
        'settings': {
            'project_sql_id': 58826,
            'userlink': {
                'url': '/',
                'text': 'Request APCD Access'
            },
            'description': "This step confirms that you have access to the APCD Portal"
        },
        'retry': True
    },
        ...
    ]
```

3. Try to log in. You should be prompted to Request Access. The link in this configuration will take you to the site root '/'
4. Refresh the page. You should see it attempt to re-process the APCD membership check
5. Change the project_sql_id to `57734`
6. Shut down your containers with `make stop` and then start them up again with `make start`. 
7. Refresh your onboarding page. This time, (assuming you are on the APCD project) the apcd membership check will succeed and continue with the rest of onboarding

## UI

![image](https://user-images.githubusercontent.com/17181582/182952754-d4ac1b0a-cf2f-4807-8d5c-238d36f0b8e9.png)

## Notes
